### PR TITLE
Faster reverse!(::BitVector)

### PR DESCRIPTION
**Pull request by dcjones**
This is an implementation bit vector reversal that's seems to be about 2x as fast as the current one:

``` julia
function f()
    srand(1234)
    for _ in 1:1000
        b = bitrand(rand(1:10000000))
        reverse!(b)
    end
end
f()
@time f()
```

```
   2.327 seconds      (38417 k allocations: 1187 MB, 2.98% gc time)
```

``` julia
function g()
    srand(1234)
    for _ in 1:1000
        b = bitrand(rand(1:10000000))
        fast_reverse!(b)
    end
end
g()
@time g()
```

```
   1.130 seconds      (9979 allocations: 601 MB, 4.41% gc time)
```

### CODE REVIEW

**Comment by yuyichao** [comment]
I'm wondering how much time is actually spent on random number generator.

**Comment by ScottPJones** [comment]
Yes, I saw that problem in the C microbenchmarks... they were really testing the random number generator, not what supposedly was being benchmarked...
@dcjones I'll bet you get much much better numbers if you pull the randomization out of the benchmark.

**Comment by yuyichao** [comment]
And I usually deal with different input sizes by benchmarking different sizes manually. It also helps to see if how the speedup depending on the input.

**Comment by ScottPJones** [comment]
@yuyichao Good point... for my string conversions, I benchmarked from 8 characters up to 4M characters... (increasing size ^2 each time), seemed to work well.

**Comment by dcjones** [comment]
Good points, here's a slightly more careful one:

```julia
function f()
    for n in [1, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000, 1000000000]
        b = bitrand(n)
        println("n = ", n)
        @time reverse!(b)
        @time fast_reverse!(b)
    end
end
```

```
n = 1
   1.387 microseconds (1 allocation: 80 bytes)
     751 nanoseconds
n = 100
   9.957 microseconds (2 allocations: 96 bytes)
     429 nanoseconds
n = 1000
   1.248 microseconds (9 allocations: 208 bytes)
     624 nanoseconds
n = 10000
   4.964 microseconds (79 allocations: 1328 bytes)
   2.089 microseconds
n = 100000
  53.090 microseconds (782 allocations: 12576 bytes)
  17.172 microseconds
n = 1000000
 378.419 microseconds (1 allocation: 80 bytes)
 168.150 microseconds
n = 10000000
   3.717 milliseconds (1 allocation: 80 bytes)
   1.747 milliseconds
n = 100000000
  36.211 milliseconds (1 allocation: 80 bytes)
  18.643 milliseconds
n = 1000000000
 365.744 milliseconds (1 allocation: 80 bytes)
 165.768 milliseconds
```

I get about the same if I reverse the calls, so that's not conflating it much.

**Comment by ScottPJones** [comment]
Since this is dependent on n mod 64, maybe you should have tests for all values of n mod 64 also, at different sizes, so n = 0 + 1..64, n = 256 + 1..64, n = 65536 + 1..64, ...

**Comment by dcjones** [comment]
```julia
function f()
    n = 2^30
    for i in 0:63
        b = bitrand(n + i)
        println("n = ", n + i)
        @time reverse!(b)
        @time fast_reverse!(b)
        @assert fast_reverse(b) == reverse(b)
    end
end
```

I'll spare you the full output. This version does 170-190 milliseconds on every one, the one currently in base is 520-580 milliseconds on all except `n % 64 == 0`, where it's 370 milliseconds.

 `n %  64 == 0` could be special cased in this version and made faster, but I didn't think it would be worth it.

**Comment by carlobaldassi** [comment]
Very nice optimization! Travis complains about trailing white space.

File: base/bitarray.jl [hunk 1]
```diff
@@ -1146,31 +1146,61 @@ function reverse_bits(src::UInt64)
...
 
-    for i = 1:hnc
-        j = ((i - 1) << 6)
-        aux_chunks[1] = reverse_bits(B.chunks[i])
-        copy_chunks!(B.chunks, j+1, B.chunks, n-63-j, 64)
-        B.chunks[i] = reverse_bits(B.chunks[i])
-        copy_chunks!(B.chunks, n-63-j, aux_chunks, 1, 64)
+    k = _mod64(n)
+    h = 64 - k
+    if k == 0
```
Comment (carlobaldassi) [inline base/bitarray.jl]: You can do `k = _mod64(n+63) + 1` to avoid this check.

Comment (dcjones) [inline base/bitarray.jl]: Thanks for the tip!

**Comment by ScottPJones** [comment]
:+1: very nice speedup!

**Comment by ScottPJones** [comment]
Did you only test this time at 2^30 + 0:63?  The main size should probably be varied also, for maybe 8 or so different sizes, just to prove there's no regression, with n = 1:64...

**Comment by StefanKarpinski** [comment]
I think it's pretty clear that this is a significant improvement. There's no suggestion that there are nefarious cases where it's not better. +1

**Comment by ScottPJones** [comment]
It's not a matter of believing that there are `nefarious` cases, however a lot of algorithms fall down with small n (because of extra, fixed overhead), so it is important that that be tested as well.
Things need to be rigorously tested, @StefanKarpinski, otherwise performance regressions can slip in.

**Comment by dcjones** [comment]
```julia
function f()
    slow_count = 0
    max_slowness = 0.0
    max_slowness_n = 0

    for k in 1:30
        for i in 0:63
            n = 2^k + i
            b = bitrand(n)

            t0 = time_ns()
            reverse!(b)
            t_base = time_ns() - t0

            t0 = time_ns()
            fast_reverse!(b)
            t_fast = time_ns() - t0

            if t_fast > t_base
                max_slowness = max(max_slowness, t_fast / t_base)
                max_slowness_n = n
                slow_count += 1
            end

            @assert fast_reverse(b) == reverse(b)
        end
    end

    @printf("slower in %0.2f%% of cases\n", 100.0 * slow_count / (30 * 64))
    @printf("maximum slowness %0.2f%% with n = %d\n", 100.0 * max_slowness, max_slowness_n)
end
```

```
slower in 0.10% of cases
maximum slowness 104.98% with n = 60
```

Convinced?

**Comment by yuyichao** [comment]
I'm surprised that it is actually slower in 0.1% of the case in your test. It must have been a huge outlier to overcome the overhead of allocationg....

**Comment by dcjones** [comment]
Well all those 0.1% of cases occur with very small n's, in which case both functions are fast enough that the measurements are pretty noisy.

**Comment by ScottPJones** [comment]
@dcjones It wasn't a matter of me being convinced or not, I already said I thought it was a very nice change, I just believe in doing very thorough testing (esp. after having had some very bad experiences last year when that didn't happen [not from my changes!]).  Much better to assure yourself and everybody else now that there aren't any strange regressions for any sizes, than have something bite you and make it need to be reverted or fixed quickly later on...
So, LGTM!

**Comment by tkelman** [comment]
I especially like the diagram in the comments.

**Comment by StefanKarpinski** [comment]
Yeah, the diagram is a nice, and very [**@dcjonesian**](https://github.com/dcjones) touch.

**Pull request merged**